### PR TITLE
[DataFrame] Implement df.merge

### PR DESCRIPTION
**Pull request by devin-petersohn**
Implement the `DataFrame.merge` method that does a Database style join on common columns. Depends on #1937.

TODO:

- [x] Add tests for merge

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -480,22 +484,31 @@ def _update_inplace(self, row_partitions=None, col_partitions=None,
...
-            # lengths and indices
-            self._row_metadata = _IndexMetadata(self._block_partitions[:, 0],
-                                                index=index, axis=0)
-            self._col_metadata = _IndexMetadata(self._block_partitions[0, :],
-                                                index=columns, axis=1)
+        if col_metadata is not None:
+            self._col_metadata = col_metadata
+        else:
+            assert columns is not None, \
+                "Columns must be passed without col_metadata"
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: "If col_metadata is None, columns must be passed in"

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -2281,9 +2431,100 @@ def merge(self, right, how='inner', on=None, left_on=None, right_on=None,
...
+                              right._row_metadata._lengths) or right_index:
+
+            repartitioned_right = np.array([_match_partitioning._submit(
+                args=(df, self._row_metadata._lengths, right.index),
+                num_return_vals=len(self._row_metadata._lengths))
+                for df in right._col_partitions]).T
+        else:
+            repartitioned_right = right._block_partitions
+
+        left_cols = ray.put(self.columns)
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Put this before column merge

**Review by p-yang — CHANGES_REQUESTED** [review]

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -2454,15 +2455,67 @@ def merge(self, right, how='inner', on=None, left_on=None, right_on=None,
...
+
         # This can be put in a remote function because we don't need it until
         # the end, and the columns can be built asynchronously. This takes the
         # columns defining off the critical path and speeds up the overall
         # merge.
-        new_columns = _merge_columns.remote(self.columns, right.columns, *args)
+        new_columns = _merge_columns.remote(left_cols, right_cols, *args)
 
-        # TODO ERROR CHECK
+        if on is not None:
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Perform error checking before defining `new_columns` so users don't get a local error followed by a remote error.

**Review by robertnishihara — APPROVED** [review]

**Comment by devin-petersohn** [comment]
Jenkins, test this please

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Pull request merged**